README.md
# Vera Cloud De-coupler/Re-coupler

**CURRENT VERSION: 23056**

**KNOW WHAT YOU'RE GETTING INTO: READ THIS ENTIRE DOCUMENT BEFORE DOING ANYTHING.**

These scripts decouple a Vera Home Automation Controller from the many cloud services to which it normally connects. It is generally accepted that Veras are more stable when not connected to the mother ship, as outages of these cloud services often causes spurious reloads of Luup or even reboots of the system. Decoupled systems use less RAM and CPU as well.
Decoupling also improves system reliability when Internet access is not available.

The only required services are DNS and NTP (time), and a lapse in the former does not affect system stability. If one provides a stable local NTP server (which can be as simple and inexpensive as a RPi/ESP8266/ESP32 with an add-on RTC), then it is possible to run a decoupled Vera entirely off-grid with no Internet access whatsoever and have a stable system.

The scripts also make facilities for local (LAN) storage of rotated log files and daily configuration backups.

What decoupling *does not do* is restrict the Vera from accessing cloud or other Internet-based services at all. While it principally removes Vera's cloud services, it does not prevent you from using other services if they are important to your automations/configuration. A decoupled Vera will therefore happily collect weather data from a third-party API, or operate your garden sprinklers, smart garage door openers or thermostats, through their respective cloud services. Eliminating these cloud connections, if that is your goal, is elective for you and simply a matter of not using those plugins.

## System Requirements

These scripts are intended only for Vera Plus, Secure and Edge system running 7.0.29 firmware or higher (to the current release version as of this writing). Users on 7.0.30, which existed only in a public beta, are advised to upgrade to 7.0.31, as the earlier firmware has a specific defect that causes "Christmas Lights" (blinking status lights on an otherwise dead unit) when Internet access becomes unavailable for an extended period of time.

Do not use these scripts on any Vera that does not meet the above requirements.

## Cautions and Disclaimer

**THESE SCRIPTS ARE EXPERIMENTAL. YOUR USE OF THESE SCRIPTS IS SOLELY AND ENTIRELY AT YOUR OWN RISK, WITHOUT RESERVATION.** I've received no help from Vera in creating these scripts, it's all based on years of discovery and learning about these systems organically. But I have no direct visibility to the source code or internals of many Vera-created parts of these systems, including Luup itself and the many tools they've added to the base OS to make it their own. Most of the learning has therefore been done in a wake of experiments, many failed; some of those failures are obvious immediately, and some take weeks or months to manifest. Things can change on these systems without notice, so while these appear to work now, Vera/eZLO can do anything it wants, including taking specific action to prevent the operation of these scripts, in future. **THESE SCRIPTS ARE THEREFORE OFFERED AS-IS AND AS-AVAILABLE WITH NO WARRANTIES, EXPRESS OR IMPLIED, WHATSOEVER.**

Decoupling from Vera's cloud is not without consequences; these include:

* Remote access isn't possible, including that by Vera/eZLO support, unless you later recouple. **Vera Support will likely take the view that this script voids your system warranty.**
* Uploading of logs to their cloud services doesn't happen, so even if you recouple and give Vera Support access to your unit, they won't have logs for events prior to the recouple.
* Certain pages within the UI will not work in the absence of Vera's cloud services, including *Settings > Location* and *Settings > Firmware*.
* Native Vera notifications, including those driven through Reactor and VeraAlerts, will not work. Other notification methods should continue to be available, however.
* You will not have access to the App Marketplace for plugins (you can still manually install them via *Apps > Develop apps*).
* Daily backups of your system to Vera's storage won't be done. The script provides a facility for daily backups to be performed and uploaded to a local server that you provide, if you wish. Otherwise, you will need to do backups manually.

Common sense applies here. Vera has gone to great lengths to design a system that runs tightly with their cloud services. Disabling their cloud services forces the system to run in a manner counter to its design, and that has inherent risks known, anticipated, and unforeseen. You accept these risks as your own and are going in eyes wide open.

## Preparation

First, make sure your Vera is fully configured:

* The unit is registered and listed in your Controllers list at https://home.getvera.com;
* The unit's location settings are properly set (*Settings > Location*);
* The unit is running the version of firmware that you want to run, and that version is supported by this tool;
* All of the other settings under *Users & Account Info* are properly configured;
* You have installed any plugins from the App Marketplace that can only be installed from there.

Second, you're going to need to decide if you want to continue to use external (cloud) time servers, or use a local (LAN) time server. It's not strictly necessary to use a local LAN server, and there are many ways to build such a server and end up no better than if you didn't have one and just stuck with the cloud. See my discussion toward the end of this document regarding the merits and problems of various options.

* If your Internet access is stable most of the time and you rarely have power failures, and when you do have power failures, it's OK with you if the Vera needs a few minutes and maybe an automatic reboot/reload or two before it stabilizes, then using the default cloud time servers (pool.ntp.org) is OK. Note that even brief periods of "bogus" time on your Vera can cause it to fire date/time-based automations errantly.
* If you regularly have no Internet access, frequent prolonged outages of Internet, frequent power failures, or use a lot of time-based automations, then setting up a local NTP server is likely the better choice. Again, see the discussion on choosing/building an NTP time server near the end of this document.

Finally, you'll need to make a similar decision for DNS. The default DNS servers are the Google DNS servers at 8.8.8.8 and 8.8.4.4. This is probably fine, but there are good arguments to be made that a local DNS service provided by your router or another device in your network is actually a better choice. Either way, the availability of DNS will not affect system stability, so do what makes the most sense to you (I use my router's DNS cache/forwarder).

> TIP: Your NTP and DNS servers, and your Vera, should all use static IP addresses, in case the network's DHCP server is down or fails.

## Installing the Scripts

To install the scripts, SSH into your Vera, and then:

1. In your SSH window, type `curl -L -o decouple.zip https://codeload.github.com/toggledbits/Vera-Decouple/zip/main`
2. Unzip the script package: `unzip decouple.zip`

Once you've unzipped the archive, you may delete the ZIP archive if you wish.

## Configuring for Decoupling

Configuration is done by modifying the `decouple-config.sh` file (only &mdash; do not modify the other files). That means you'll need to use an editor on your Vera, and by default, the standard editor is `vi`, Bill Joy's wonderful ubiquitous visual editor that's as easy to learn as a saxophone. If you are comfortable in `vi`, no problem; if you need a more basic editor, proceed immediately to "Running `decouple.sh`" below, and the script will offer to install the `nano` editor for you. You can then edit the configuration file and run the script again.

Everything in the configuration file has reasonable defaults. The default value will be used when the configuration variable is either commented out or set to blank. A variable is commented out when a '#' appears in the left margin before its name. To uncomment a variable, remove that '#', and then add any value to the right of the equal ('=') sign. Do not change the name of the variable, or add any spaces before or after the equal sign. Although it is usually not necessary, you can surround the value in double-quote marks (i.e. `DNSSERVER=192.168.0.1` and `DNSSERVER="192.168.0.1"` are equivalent).

OK. Ready? Let's see what we can configure. Remember, if any config variable is commented out or blank, the default shown will be used:

* `NTPSERVER`: [default: `"0.openwrt.pool.ntp.org 1.openwrt.pool.ntp.org"`] Set to the IP address(es) of the NTP servers you would like to use. To remove dependency on Internet access, this should be one or more servers on your local LAN.
* `DNSSERVER`: [default: `"8.8.8.8 8.8.4.4"` (Google DNS)] Set to the DNS servers you would like to use. The use of local or cloud DNS services makes little difference in stability to your Vera system. If your router provides DNS cache/forwarder service, that's a fine choice.
* `KEEP_MIOS_WEATHER`: [default: 0] If non-zero, the MiOS cloud weather service will not be decoupled/disabled. This is for users of the VOTS plugin.
* `LOG_SERVER`: [default: no log storage] Set to the (one) IP address of the server to receive logs. If blank/unset, logs are discarded. If set, you will also need to set `LOG_USER` and `LOG_PASS` to the username and password, respectively, of the FTP account on that server to receive the log files. You must also create a subdirectory of that account's home directory with the same name as the serial number of your Vera unit. The logs will be uploaded to this directory. FTP is the only protocol supported by this proces.
* `SYSLOG_SERVER`: [default: no syslog logging] Set to the IP address of a SysLog remote server to receive logging from the Vera. If blank/not set, no remote Syslog logging will occur. The default protocol for logging is UDP, on destination port 514. You may the protocol to TCP by setting `SYSLOG_PROTO=tcp`; the port can be changed by setting `SYSLOG_PORT`.
* `DAILY_BACKUP_SERVER`: [default: no automatic daily backups] Set to the IP address of a server to receive daily configuration backups from your Vera system. When you decouple from the cloud, the daily backups cannot be stored on Vera's servers. If blank/not set, no daily backups will be done and you must back up manually. You can set `DAILY_BACKUP_PROTO` to one of `ftp` (the default), `ftps` (for FTP+SSL), or `scp`. You should also set `DAILY_BACKUP_USER` and `DAILY_BACKUP_PASS` to the username and password of the account to receive the backup archives. A subdirectory in that account's home of the same name as the Vera serial number is required as well. Password-less backup using public key authentication is possible; see below.

> If you are using `scp` for daily local backups and would like to use public key authentication rather than hard-coding a password here, set `DAILY_BACKUP_PASS` to `@`. Then run `decouple.sh`. After it runs, you will find a file `vera_ssh_key.txt` in the `/root` folder. Append the contents of this file to the `authorized_keys` file or equivalent for the *target system's* user account (e.g. on Linux, this would be `~user/.ssh/authorized_keys` on the target server, where `user` is the username you assigned to `DAILY_BACKUP_USER`; check your Linux system's `ssh` documentation for permissions required on the `.ssh` subdirectory and `authorized_keys` files if they don't already exist and you need to create them).

## Running `decouple.sh`

**ALWAYS BACK UP BEFORE RUNNING THE SCRIPT. BACK UP YOUR ZWAVE NETWORK AS WELL.**

After making the necessary changes to the `decouple-config.sh` file, you can run the script: `sh decouple.sh`. The script will perform the necessary steps, and save original copies of modified configuration files in a (hidden) save directory, in case you want to recouple later. Follow any instructions the script gives you. Normally, the only instruction you will get is to reboot at the end. Once you've rebooted, decoupling is in full effect.

```
root@MiOS_500xxxxx:~# sh decouple.sh
Running decouple.sh version 20288
Preliminary checks OK. Decouple this Vera from cloud services [Y/N]? y
Setting NTP server(s) to 192.168.0.15 192.168.0.44...
Setting DNS server(s) to 192.168.0.15 192.168.0.44...
Enabling syslog remote logging to 192.168.0.15...
Setting up daily system backups to 192.168.0.164 via ftp...
Disabling remote access...
Disabling NetworkMonitor...
Updating root's crontab...
Decoupling Vera cloud servers...
Turning on log uploads to 192.168.0.15...
Connecting log server 192.168.0.15 for upload/rotation...

OK. We're done here. Your Vera has been decoupled from the cloud services.
When ready, please reboot by running /sbin/reboot
root@MiOS_500xxxxx:~# /sbin/reboot
    <remote server has disconnected>
```

> It's normal and expected that a decoupled Vera will have its "Service" LED extinguished. This indicates that the system is not connected to the Vera/MiOS/eZLO cloud services.

The default location for the decoupler's saved configuration `/root/.decouple-saved` (you may need to do `ls -a` to see it), and I don't recommend that you change it, and don't delete it. It's not a bad idea to `tar` it along with the decouple and recouple scripts (e.g. `tar czf decouple-saved.taz .decouple-saved decouple.sh recouple.sh`), and save the archive off-system.

If something goes wrong, you can usually fix the problem (following message instructions or enlisting my help &mdash; see *Questions and Support* below), and then safely re-run the script. The script has considerable safeguards to ensure that original configurations are captured and preserved, and modifications are done properly if they haven't already been done.

If you decide you want to change your decoupling configuration, it is usually only necessary to modify the `decouple-config.sh` file with your changes and re-run `decouple.sh`. It is *not* necessary to recouple first.

## Recoupling to the Mother Ship

The `recouple.sh` script is provided to allow you to recouple your Vera to the company cloud services.

> Updating firmware, doing a factory reset, or restoring a backup made before decoupling will also recouple your Vera.

Once you've recoupled, the decouple save directory (default `/root/.decouple-saved`) should be regarded as invalid and discarded.

If you lose the decouple save directory or it becomes corrupt, the script will warn you and stop. You can override the stop by supplying the `-f` option on the command line. Any missing configuration will simply be restored to factory default, and the Vera should eventually update it to whatever settings the cloud services determine are correct for your unit. This is a fail-safe to ensure that you can always get back to cloud-connected service.

## Questions and Support

**THIS IS VERY IMPORTANT! PLEASE PLAY BY THE RULES!**

Because discussion of decoupling has in the past drifted into territory that has gotten people banned from the Vera Community, please **do not** ask questions or report problems there. Use the [Issues](https://github.com/toggledbits/Vera-Decouple/issues) section of this repository for that purpose exclusively.

## Considerations for Your Own NTP Time Server

In the Vera world, the quality of a time source is based on availability more than accuracy. If the time server is unavailable when Vera boots, Vera will boot up with a default time that is inaccurate, perhaps unacceptably so, and this will persist until the time server can be reached. During this period, date and time-based automations will run errantly because the system time is just wrong. When Vera finally can reach a time server, it often reboots, since large adjustments of time cause bigger problems than just rebooting.

> Decoupled Vera systems do not go through a bunch of usually-useless gyrations and attempts to set the world right that coupled systems do when time or Internet access isn't available. This adds considerable stability to the system, and reduces the problem to just a boot-time issue.

The time problem is solved by building a highly-available, sufficiently-accurate time source on your LAN. This sounds a bit daunting, but it can actually be done for very little money and achieve excellent results. And there's a spectrum of solutions, so what you do, how difficult it is to implement and maintain, and how much it costs really depends on how far you want to go minimizing the problem.

A few options look like this:

* Best: time server has a hardware real time clock synced to external sources and both the clock and the entire system are battery-backed. Battery-backup of the RTC itself usually takes the form of a coin-cell battery. System backup is usually a traditional UPS or similar, larger battery system. The advantages here compound: the system is immune to brief power failures and will thus provide time service immediately when power is restored and other network devices/systems restart; if the power outage exceeds the UPS capacity, the hardware real time clock is still able to keep time, and when the time server reboots, it will begin with a query to the RTC to get a time very close to actual right from the start; synchronization with external sources discliplines the clock for best accuracy/consistency, and when synchronization isn't possible (e.g. Internet outage), the software and hardware-backed system clock continues to run. Most desktop PCs and full-size server PCs have RTCs, but be wary &mdash; this is one of the areas where manufacturers engineer cost out of the system quite often.
* Good: time server without a hardware RTC, but running an NTP service and externally synced, and battery-backed (UPS or similar). The expectation here is that the system stays running in all but the worst of cases (prolonged power outage). Because the system stays running, it is usually immediately available to answer time queries when other devices have power restored, increasing the likelihood that Vera, in particular, will not have to start with a bogus default time and resync later (it can resync immediately). During an Internet outage, the system can't sync, but the software clock is free-running and should be sufficiently accurate, as long as the time server itself doesn't reboot. If that happens, all bets are off.
* Acceptable: time server with a hardware RTC and no external synchronization, UPS-backed. A network with no Internet access would use this approach, for example. The network time will at least be consistent across all devices, but the time will only be as accurate as the hardware clock can achieve, and you'll probably need to adjust it manually from time to time.
* Unacceptable -- Time server with no hardware real time clock and no UPS/battery backup. This is unacceptable because there is no way for the time server to accurate time when it starts up (particularly if there is no Internet access), and the time server itself may not boot up sufficiently quickly to respond to the Vera's queries during its boot-up, resulting in the Vera also having bogus default time at startup.

If you use your router for NTP service on your network, and your router is not plugged into a UPS, you are in the *unacceptable* category unless you can confirm that your router has a hardware RTC (if it does, you're *good* or *acceptable* depending on system configuration).

If you use your NAS connected to a UPS for NTP service, you are in the *good* category if it has no real time clock, or *best* if it does.

If you use an RPi/ESP8266/ESP32 with an RTC hat/shield/board, the RTC is battery-backed, and the RPi itself is UPS-powered, you are in the *best* category. This can usually be achieved very cheaply and is highly recommended. If you can't do this right from the start, make it a plan to get there. It'll be fun and educational. For myself, I've chosen an Adafruit Ethernet Featherwing ($19.95) and a Feather RTC ($8.95); I'll publish the code for my solution in future, and I may also do a GPS version (which gives stratum 1 time service).


CHANGELOG.md
# Vera-Decouple ChangeLog

https://github.com/toggledbits/Vera-Decouple

## 23056

* Switch key for password-less backups to Vera's ECDSA key (was RSA). The RSA key type is no longer accepted by modern SSH servers (such as that on Ubuntu 22.04 and up) because of the weak SHA1 hash it uses. Since Veras are now quite old in the software sense, the encryption/key choices are few, and as of now, ECDCSA is the only remaining key type that works with newer SSH servers. If you previously configured password-less backup, you will need to redo the steps described in the README to add the new ECDSA key.

## 21222

* Support for 7.32 beta 4 (5385/6/7)
* Fix error in command line argument handling in recoupler.

## 20347

* New KEEP_MIOS_WEATHER flag to prevent decouple/disable of weather cloud service. This is for users of the VOTS plugin and was suggested by JW (issue #3).

## 20301

* Disable apps.mios.com

## 20294

* Make sure the alerts stay clear and system doesn't get bogged down trying to send alerts that cannot be delivered.

## 20290

* Fixed an issue where 20289 inserted an additiona  line feed in a cron job. To repair, you can either hand-edit the cron script (`crontab -u root -e`, join the "23 0" job and the line after), or just (re)run `decouple.sh` after updating to this version.

## 20289

* Initial release for Plus and Edge. Secure should be the same as Plus, but it needs to be tested.

decouple.sh
#!/bin/sh

# ------------------------------------------------------------------------------
#
# decouple.sh -- Shell script to decouple Vera from its cloud services.
# Copyright (C) 2020,2021 Patrick H. Rigney (rigpapa), All Rights Reserved
#
# Please see decouple-config.sh for special configuration information.
# READ THE DOCUMENTATION AT https://github.com/toggledbits/Vera-Decouple
#
# ------------------------------------------------------------------------------

_VERSION=23056

askyn() {
    local __ans
    local __resultvar
    __resultvar="${1:-ans}"
    while true; do
        echo -e -n "${2:-}"
        read __ans
        case "$__ans" in
            [Yy]* )
                eval "${__resultvar}=Y"
                break
                ;;
            [Nn]* )
                eval "${__resultvar}=N"
                break
                ;;
        esac
        echo "Please answer Y or N."
    done
}

echo "Running decouple.sh version $_VERSION"

[ -d /etc/cmh ] || { echo "$0: runs on Vera Plus, Secure, Edge only"; exit 255; }

if [ -s /proc/diag/model ]; then
    p=$(cat /proc/diag/model)
else
    p=$(awk '/^machine/ { print $4 }' /proc/cpuinfo)
fi
if [ -z "$p" ]; then
    p=$(cat /etc/cmh/vera_model)
fi
case "$p" in
    "NA301"|"G450"|"G550") # Vera Edge|Plus|Secure
        ;;
    "930") # Plus, Edge via /etc/cmh/vera_model
        ;;
    *)
        echo "$0 does not work on this Vera model (${p})."
        echo "It only works on Vera Plus, Secure, and Edge."
        exit 255
esac
fw=$(fgrep 1.7. /etc/cmh/version | sed 's/1.7.//')
if [[ "$fw" == "" || $fw -lt 4452 || $fw -gt 5387 ]]; then
    echo "$0 has not been tested on systems running $(cat /etc/cmh/version)."
    exit 255
fi
if [ ! -f decouple-config.sh ]; then
    echo "$0: can't find config file decouple-config.sh"
    exit 255
fi

if [ ! -f /usr/bin/nano ]; then
    echo "It appears the 'nano' editor is not installed. If you're comfortable with"
    echo "'vi', that's not a problem, but I can install nano for you now, if you wish."
    askyn inst_nano "Install nano editor now [y/n]? "
    if [ "${inst_nano}" == "Y" ]; then
        opkg update
        opkg install nano
    fi
fi

[ -f decouple-config.sh ] && . decouple-config.sh

SAVEDIR=${SAVEDIR:-/root/.decouple-saved}

if [ -f ${SAVEDIR}/recoupled ]; then
    rd=$(stat -c %Y ${SAVEDIR}/recoupled)
    dd=$(stat -c %Y ${SAVEDIR}/servers.conf)
    nd=$(date +%s)
    rdiff=$(expr $(expr $nd - $rd) / 86400)
    dver=$(head -1 ${SAVEDIR}/decouple-version)
    recommend=2
    [ $rdiff -gt 30 ] && recommend=1
    [ $dver -lt $_VERSION ] && recommend=1
    cat <<RECOUPLECHECK

The decouple save directory appears to have already been used to recouple. If
it has been some time since your last decouple/recouple, or you decoupled with
a much earlier version of this project, the saved configuration could be stale
and invalid.

    1) CLEAR the previously-saved data and start fresh (recommended if it has
       been a long time since you recoupled (e.g. more than a few months).
    2) KEEP the previously-saved data (recommended if you just recently re-
       coupled and are now re-decoupling).
    3) Exit now; no changes will be made.

Decouple data saved on $( date -d @${dd} ) using version $(head -1 ${SAVEDIR}/decouple-version)
     Recoupled last on $( date -d @${rd} ) $rdiff days ago
Recommended answer: ${recommend}. If you're not sure, "2" (KEEP) is the safest option.
RECOUPLECHECK
    while true; do
        echo -n "Choose 1, 2 or 3: "
        read ans
        case "$ans" in
            [123] )
                break
                ;;
        esac
    done
    [ "$ans" == "3" ] && { echo "Exiting without changes."; exit 255; }
    [ "$ans" == "1" ] && rm -f ${SAVEDIR}/*
    # recoupled file is removed later (when decouple succeeds)
    echo ""
fi

askyn ans "Ready to go. Decouple this unit from Vera/eZLO cloud services [Y/N]? "
[ "${ans:-}" == "Y" ] || { echo "Exiting without changes."; exit 255; }

mkdir -p ${SAVEDIR}/

if [ -n "${NTPSERVER:-}" ]; then
    echo "Setting NTP server(s) to ${NTPSERVER}... "
    words="${NTPSERVER}"
else
    echo "Setting default NTP servers (openwrt.pool.ntp.org)..."
    words="0.openwrt.pool.ntp.org 1.openwrt.pool.ntp.org"
fi
if [ -n "$words" ]; then
    uci delete system.ntp.server
    while uci delete ntpclient.@ntpserver[-1] >/dev/null 2>&1; do n=0; done
    for s in $words; do
        key=$(uci add ntpclient ntpserver)
        uci set ntpclient.$key.hostname=$s
        uci set ntpclient.$key.port=123
        uci add_list system.ntp.server=$s
    done
    uci commit ntpclient
    uci commit system.ntp
    /etc/init.d/sysntpd restart
    /etc/init.d/ntpclient restart
else
    echo "*** No NTP configuration changes are being made."
fi

if [ -n "${DNSSERVER:-}" ]; then
    echo "Setting DNS server(s) to ${DNSSERVER}..."
    words="${DNSSERVER}"
else
    echo "Setting default DNS servers (Google DNS)..."
    words="8.8.8.8 8.8.4.4"
fi
if [ -n "$words" ]; then
    uci delete dhcp.@dnsmasq[0].server
    for s in $words; do
        uci add_list dhcp.@dnsmasq[0].server="$s"
    done
    if [ ${KEEP_MIOS_WEATHER:-0} -ne 0 ]; then
        noms="apps.mios.com"
    else
        noms="weather.mios.com apps.mios.com"
    fi
    for s in ${noms}; do
        uci add_list dhcp.@dnsmasq[0].server="/$s/"
    done
    uci commit dhcp
    /etc/init.d/dnsmasq restart
else
    echo "*** No DNS configuration changes are being made."
fi

if [ -n "${SYSLOG_SERVER:-}" ]; then
    echo "Enabling syslog remote logging to ${SYSLOG_SERVER}..."
    uci set system.@system[0].log_ip=${SYSLOG_SERVER}
    uci set system.@system[0].log_port=${SYSLOG_PORT:-514}
    uci set system.@system[0].log_proto=${SYSLOG_PROTO:-udp}
    uci commit system
    /etc/init.d/log restart
elif fgrep log_ip /etc/config/system >/dev/null; then
    echo "Disabling syslog remote logging..."
    uci delete system.@system[0].log_ip >/dev/null 2>&1
    uci delete system.@system[0].log_port >/dev/null 2>&1
    uci delete system.@system[0].log_proto >/dev/null 2>&1
    uci commit system
    /etc/init.d/log restart
fi

if [ -n "${DAILY_BACKUP_SERVER:-}" ]; then
    echo "Setting up daily system backups to ${DAILY_BACKUP_SERVER} via ${DAILY_BACKUP_PROTO:-ftp}..."
    usr=${DAILY_BACKUP_USER:-verabackup}
    if [ "${DAILY_BACKUP_PROTO:-}" == "scp" ]; then
        if [ "${DAILY_BACKUP_PASS:-}" == "@" ]; then
            opts="-i /etc/dropbear/dropbear_ecdsa_host_key"
            targ="${usr}@${DAILY_BACKUP_SERVER}"
        else
            opts=""
            targ="${usr}:${DAILY_BACKUP_PASS:-}@{DAILY_BACKUP_SERVER}"
        fi
        cat <<SCPBACKUPSCRIPT >/usr/bin/decouple_daily_backup.sh
#!/bin/sh

# DO NOT EDIT -- AUTOMATICALLY GENERATED FILE
# This file is part of rigpapa's decoupler. See https://github.com/toggledbits/Vera-Decouple
# Created by $_VERSION on $(date)

ser="\$(nvram get vera_serial)"
target="${targ}"
dest_path="\${ser}/ha-gateway-backup_\${ser}_\$(cat /etc/cmh/version)_\$(date +%F).tgz"

sh /usr/bin/backup-store.sh D
scp ${opts} /tmp/Backup_download.tgz \${target}:\${dest_path}
rm -f /tmp/Backup_download.tgz
SCPBACKUPSCRIPT
    else
        cmd="curl -T - -m 300 --connect-timeout 15"
        if [ "${DAILY_BACKUP_PROTO:-}" == "ftps" ]; then
            cmd="${cmd} --ftp-ssl"
        elif [ "${DAILY_BACKUP_PROTO:-ftp}" != "ftp" ]; then
            echo "Invalid DAILY_BACKUP_PROTO ('${DAILY_BACKUP_PROTO}'); only ftp, ftps, scp allowed."
            exit 255
        fi
        cat <<FTPBACKUPSCRIPT >/usr/bin/decouple_daily_backup.sh
#!/bin/sh

# DO NOT EDIT -- AUTOMATICALLY GENERATED FILE
# This file is part of rigpapa's decoupler. See https://github.com/toggledbits/Vera-Decouple
# Created by $_VERSION on $(date)

ser="\$(nvram get vera_serial)"
target_host="${DAILY_BACKUP_SERVER}"
target_auth="${usr}:${DAILY_BACKUP_PASS:-}"
dest_path="\${ser}/ha-gateway-backup_\${ser}_\$(cat /etc/cmh/version)_\$(date +%F).tgz"

sh /usr/bin/backup-store.sh D
cat /tmp/Backup_download.tgz | ${cmd} ftp://\${target_auth}@\${target_host}/\${dest_path}
rm -f /tmp/Backup_download.tgz
FTPBACKUPSCRIPT
    fi
    chmod +rx /usr/bin/decouple_daily_backup.sh
fi

if [ ! -f ${SAVEDIR}/services.conf ]; then
    cp /etc/cmh/services.conf ${SAVEDIR}/services.conf || exit 1
fi
if ! fgrep 'Permissions_Relay=0' /etc/cmh/services.conf >/dev/null; then
    echo "Disabling Vera's remote access services..."
    sed -i 's/Permissions_Relay=.*/Permissions_Relay=0/' /etc/cmh/services.conf
    /etc/init.d/cmh-ra stop
fi

echo "Disabling NetworkMonitor..."
if [ -L "/usr/bin/InternetOk" ]; then
    rm -f /usr/bin/InternetOk # unlink
    echo 'exit 0 # decouple' >/usr/bin/InternetOk
fi
chmod +rx /usr/bin/InternetOk
if grep '^NM_WATCHDOG_FILE=' /usr/bin/Rotate_Logs.sh >/dev/null; then
    rm -f /usr/bin/Rotate_Logs.sh
    sed '/^NM_WATCHDOG_FILE=/,/^fi/c \
# decouple removed \
' </mios/usr/bin/Rotate_Logs.sh | sed 's/^default_Server_Log/#default_Server_Log/I' >/usr/bin/Rotate_Logs.sh
fi
chmod +rx /usr/bin/Rotate_Logs.sh
if ! grep '^exit 0 # decouple' /usr/bin/Start_NetworkMonitor.sh >/dev/null; then
    rm -f /usr/bin/Start_NetworkMonitor.sh
    sed '/===BEGIN===/a \
log "$0 disabled by decouple" # decouple \
touch /var/run/nm.stop # decouple \
exit 0 # decouple \
' </mios/usr/bin/Start_NetworkMonitor.sh >/usr/bin/Start_NetworkMonitor.sh
fi
chmod +rx /usr/bin/Start_NetworkMonitor.sh
if ! fgrep 'touch /var/run/nm.stop # decouple' /etc/init.d/check_internet >/dev/null; then
    awk '/bin\/Start_NetworkMonitor.sh/ { print "touch /var/run/nm.stop # decouple.sh"; print $0; next } { print; }' </etc/init.d/check_internet >/tmp/decouple.tmp && \
        mv /tmp/decouple.tmp /etc/init.d/check_internet
fi
chmod +rx /etc/init.d/check_internet
/etc/init.d/check_internet stop >/dev/null 2>&1
touch /var/run/nm.stop

echo "Updating root's crontab..."
if [ ! -f ${SAVEDIR}/crontab-root ]; then
    crontab -u root -l > ${SAVEDIR}/crontab-root
fi
# 20290: &servi pattern is to fix a stray newline output in 20289
crontab -u root -l | \
    grep -Fv '# decouple' | \
    grep -v '&servi *$' | \
    awk '/^#/ { print; next } /Rotate_Logs/ { print; next } { print "# "$0 }' >/tmp/decouple.tmp
cat <<CRONMORE >>/tmp/decouple.tmp
# decouple additions ($_VERSION)
0 */4 * * * rm -f /etc/cmh/alerts.json /etc/cmh/persist/* # decouple
CRONMORE
if [ -n "${DAILY_BACKUP_SERVER:-}" ]; then
    cat <<BACKUPCRON >>/tmp/decouple.tmp
23 0 * * * curl -o - -m 30 'http://127.0.0.1:3480/data_request?id=action&serviceId=urn:micasaverde-com:serviceId:ZWaveNetwork1&action=BackupDongle&DeviceNum=1' 2>&1 | logger -t decouple_daily_backup # decouple
5 0 * * * /usr/bin/decouple_daily_backup.sh 2>&1 | logger -t decouple_daily_backup # decouple
BACKUPCRON
fi
crontab -u root /tmp/decouple.tmp && rm /tmp/decouple.tmp

echo "Decoupling Vera cloud servers..."
if [ ! -f ${SAVEDIR}/servers.conf ]; then
    cp /etc/cmh/servers.conf ${SAVEDIR}/servers.conf || exit 1
fi
awk '/^Server_/ { sub("=.*$", "=127.0.0."NR); print; next } { print }' < ${SAVEDIR}/servers.conf > /etc/cmh/servers.conf
. /etc/cmh/servers.conf
for s in $(awk -F= '/^Server_/ { print $1 }' /etc/cmh/servers.conf); do
    eval "Z=\$$s"
    # echo "    + updating ${s} to ${Z}"
    nvram get mios_${s} >/dev/null && nvram set mios_${s}=${Z}
    sed -i "s/^Settings_${s}=.*/Settings_${s}=${Z}/" /etc/cmh/services.conf
done
nvram commit >/dev/null 2>&1

if [ -z "${LOG_SERVER:-}" ]; then
    echo "Turning off log uploads..."
    sed -i 's/ArchiveLogsOnServer=.*/ArchiveLogsOnServer=0/' /etc/cmh/cmh.conf
else
    echo "Turning on log uploads to ${LOG_SERVER}..."
    sed -i "s/Server_Log=.*/Server_Log=${LOG_SERVER}/" /etc/cmh/servers.conf
    sed -i "s/Server_Log_User=.*/Server_Log_User=${LOG_USER}/" /etc/cmh/servers.conf
    sed -i "s/Server_Log_Pass=.*/Server_Log_Pass=${LOG_PASS}/" /etc/cmh/servers.conf
    sed -i 's/ArchiveLogsOnServer=.*/ArchiveLogsOnServer=1/' /etc/cmh/cmh.conf
fi

# PLATFORM: NA301 on 5245, dropbear starts at 50
# PLATFORM: G450 on 5186, dropbear starts at 15 (before network at 20?)
# Start dropbear earlier, so we can debug other startup problems more easily.
if [ -L /etc/rc.d/S50dropbear ]; then
    sed -i 's/START=.*/START=21/' /etc/init.d/dropbear
    mv -f /etc/rc.d/S50dropbear /etc/rc.d/S21dropbear
fi

# Prevent the provisioning scripts from running at boot. We're provisioned,
# and running it with no auth servers stalls the boot.
rm -f /etc/rc.d/S*-provision_vera*
rm -f /etc/rc.d/S*-cmh-ra

# OpenWRT has a default sysfixtime that runs early (first) and does a good job of
# approximating a usable time prior to NTP service. Vera then runs a script that
# undoes that, disastrously. Disable the Vera script on systems that don't have RTC
# (I'm not even sure what, if any, do).
[ -e /dev/rtc0 ] || rm -f /etc/rc.d/S*mios_fix_time*

# Our own startup script
cat <<BOOTSCRIPT >/etc/init.d/decouple
#!/bin/sh /etc/rc.common

# Inserted by decouple $_VERSION

START=999

boot() {
    /usr/bin/set_led.sh off service
}
BOOTSCRIPT
chmod +rx /etc/init.d/decouple
( cd /etc/rc.d/ ; ln -sf ../init.d/decouple S999decouple )

cp decouple-config.sh ${SAVEDIR}/
echo -n "$_VERSION" >${SAVEDIR}/decouple-version

if [ -n "${DAILY_BACKUP_SERVER}" -a "${DAILY_BACKUP_PROTO}" == "scp" -a "${DAILY_BACKUP_PASS}" == "@" ]; then
    echo ; echo "*** Don't forget to append the following public key to the SSH authorized_keys"
    echo "    file for user ${DAILY_BACKUP_USER:-verabackup} on your backup server ${DAILY_BACKUP_SERVER}."
    echo "    Although it may wrap on your screen, it should be placed as a single (long)"
    echo "    line in the authorized_hosts file. A copy of this key has been left in file"
    echo "    /root/vera-ssh-pubkey.txt as well."
    dropbearkey -y -f /etc/dropbear/dropbear_ecdsa_host_key | grep -i '^ecdsa' | tee /root/vera-ssh-pubkey.txt
fi

# Remove the recoupled flag from SAVEDIR
rm -f ${SAVEDIR}/recoupled

cat <<EOF

OK. We're done here. Your Vera has been decoupled from the Vera/eZLO cloud services.
When ready, please reboot by running /sbin/reboot
EOF

exit 0
